browser/ntp_sponsored_images/android/ntp_sponsored_images_bridge.cc

```diff
@@ -0,0 +1,150 @@
+using ntp_sponsored_images::NTPSponsoredImagesService;
+using ntp_sponsored_images::ViewCounterService;
+using ntp_sponsored_images::ViewCounterServiceFactory;
+
+namespace ntp_sponsored_images {
+
+NTPSponsoredImagesBridgeFactory::NTPSponsoredImagesBridgeFactory()
+    : BrowserContextKeyedServiceFactory(
+          "NTPSponsoredImagesBridge",
+          BrowserContextDependencyManager::GetInstance()) {}
+NTPSponsoredImagesBridgeFactory::~NTPSponsoredImagesBridgeFactory() {}
+
+// static
+NTPSponsoredImagesBridge*
+NTPSponsoredImagesBridgeFactory::GetForProfile(Profile* profile) {
+  return static_cast<NTPSponsoredImagesBridge*>(
+      GetInstance()->GetServiceForBrowserContext(profile, true));
+}
+
+// static
+NTPSponsoredImagesBridgeFactory*
+NTPSponsoredImagesBridgeFactory::GetInstance() {
+  return base::Singleton<NTPSponsoredImagesBridgeFactory>::get();
+}
+
+KeyedService* NTPSponsoredImagesBridgeFactory::BuildServiceInstanceFor(
+      content::BrowserContext* context) const {
+  return new NTPSponsoredImagesBridge(Profile::FromBrowserContext(context));
+}
+
+bool
+NTPSponsoredImagesBridgeFactory::ServiceIsCreatedWithBrowserContext() const {
+  return true;
+}
+
+}  // namespace
+
+NTPSponsoredImagesBridge::NTPSponsoredImagesBridge(Profile* profile)
+    : profile_(profile),
+      view_counter_service_(ViewCounterServiceFactory::GetForProfile(profile_)),
+      sponsored_images_service_(
+          g_brave_browser_process->ntp_sponsored_images_service()) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+
+  java_object_.Reset(Java_NTPSponsoredImagesBridge_create(
+      AttachCurrentThread(), reinterpret_cast<intptr_t>(this)));
+
+  sponsored_images_service_->AddObserver(this);
+}
+
+NTPSponsoredImagesBridge::~NTPSponsoredImagesBridge() {
+  sponsored_images_service_->RemoveObserver(this);
+  Java_NTPSponsoredImagesBridge_destroy(AttachCurrentThread(), java_object_);
+}
+
+static base::android::ScopedJavaLocalRef<jobject>
+JNI_NTPSponsoredImagesBridge_GetInstance(JNIEnv* env,
+                                      const JavaParamRef<jobject>& j_profile) {
+  auto* profile = ProfileAndroid::FromProfileAndroid(j_profile);
+  return ntp_sponsored_images::NTPSponsoredImagesBridgeFactory::GetInstance()
+             ->GetForProfile(profile)->GetJavaObject();
+}
+
+base::android::ScopedJavaLocalRef<jobject>
+NTPSponsoredImagesBridge::GetJavaObject() {
+  return base::android::ScopedJavaLocalRef<jobject>(java_object_);
+}
+
+void NTPSponsoredImagesBridge::RegisterPageView(
+    JNIEnv* env, const JavaParamRef<jobject>& obj) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  if (view_counter_service_)
+    view_counter_service_->RegisterPageView();
+}
+
+base::android::ScopedJavaLocalRef<jobject>
+NTPSponsoredImagesBridge::CreateWallpaper() {
+  JNIEnv* env = AttachCurrentThread();
+
+  auto data = view_counter_service_->GetCurrentWallpaperForDisplay();
+  if (!view_counter_service_ || data.is_none())
```

Comment:
`if (!view_counter_service_ || view_counter_service_->GetCurrentWallpaperForDisplay().is_none))`?

Reply:
then we'd have to call `view_counter_service_->GetCurrentWallpaperForDisplay()` again to get `data` which seems silly

Reply:
I think it seems odd about null checking after referencing `view_counter_service_`?

Reply:
ah, I see what you're saying now. I'll fix